File: too_many_cooks_data/test/database_test.dart
/// Tests for database creation and lifecycle.
library;

import 'dart:js_interop';

import 'package:dart_node_core/dart_node_core.dart';
import 'package:nadz/nadz.dart';
import 'package:test/test.dart';
import 'package:too_many_cooks_data/too_many_cooks_data.dart';

extension type _Fs(JSObject _) implements JSObject {
  external void unlinkSync(String path);
  external bool existsSync(String path);
  external void rmdirSync(String path, _RmdirOptions options);
}

extension type _RmdirOptions._(JSObject _) implements JSObject {
  external factory _RmdirOptions({bool recursive});
}

final _Fs _fs = _Fs(requireModule('fs') as JSObject);

void _deleteIfExists(String path) {
  try {
    if (_fs.existsSync(path)) {
      _fs.unlinkSync(path);
    }
  } on Object catch (_) {}
}

void _deleteDirIfExists(String path) {
  try {
    if (_fs.existsSync(path)) {
      _fs.rmdirSync(path, _RmdirOptions(recursive: true));
    }
  } on Object catch (_) {}
}

void main() {
  test('createDb succeeds with valid path', () {
    const testDbPath = '.test_create_db.db';
    _deleteIfExists(testDbPath);

    final config = createDataConfig(dbPath: testDbPath);
    final result = createDb(config);
    expect(result, isA<Success<TooManyCooksDb, String>>());

    final db = (result as Success<TooManyCooksDb, String>).value;
    db.close();
    _deleteIfExists(testDbPath);
  });

  test('createDb creates parent directory if needed', () {
    const testDir = '.test_nested_dir';
    const testDbPath = '$testDir/subdir/data.db';
    _deleteDirIfExists(testDir);

    final config = createDataConfig(dbPath: testDbPath);
    final result = createDb(config);
    expect(result, isA<Success<TooManyCooksDb, String>>());

    final db = (result as Success<TooManyCooksDb, String>).value;
    db.close();
    _deleteDirIfExists(testDir);
  });

  test('close succeeds', () {
    const testDbPath = '.test_close.db';
    _deleteIfExists(testDbPath);

    final config = createDataConfig(dbPath: testDbPath);
    final createResult = createDb(config);
    final db = (createResult as Success<TooManyCooksDb, String>).value;

    final closeResult = db.close();
    expect(closeResult, isA<Success<void, DbError>>());

    _deleteIfExists(testDbPath);
  });

  test('schema version is set correctly', () {
    expect(schemaVersion, 1);
  });

  test('error codes are defined', () {
    expect(errNotFound, 'NOT_FOUND');
    expect(errUnauthorized, 'UNAUTHORIZED');
    expect(errLockHeld, 'LOCK_HELD');
    expect(errLockExpired, 'LOCK_EXPIRED');
    expect(errValidation, 'VALIDATION');
    expect(errDatabase, 'DATABASE');
  });
}

File: too-many-cooks/test/database_test.ts
/// Tests for database creation and lifecycle.

import { describe, it } from "node:test";
import assert from "node:assert";

import fs from "node:fs";

import {
  createDataConfig,
  createDb,
  ERR_NOT_FOUND,
  ERR_UNAUTHORIZED,
  ERR_LOCK_HELD,
  ERR_LOCK_EXPIRED,
  ERR_VALIDATION,
  ERR_DATABASE,
} from "../lib/src/data/data.js";
import { SCHEMA_VERSION } from "../lib/src/data/schema.js";

const deleteIfExists = (path: string): void => {
  try {
    if (fs.existsSync(path)) {
      fs.unlinkSync(path);
    }
  } catch {
    // ignore
  }
};

const deleteDirIfExists = (path: string): void => {
  try {
    if (fs.existsSync(path)) {
      fs.rmSync(path, { recursive: true });
    }
  } catch {
    // ignore
  }
};

describe("database", () => {
  it("createDb succeeds with valid path", () => {
    const testDbPath = ".test_create_db.db" as const;
    deleteIfExists(testDbPath);

    const config = createDataConfig({ dbPath: testDbPath });
    const result = createDb(config);
    assert.strictEqual(result.ok, true);

    if (result.ok) {
      result.value.close();
    }
    deleteIfExists(testDbPath);
  });

  it("createDb creates parent directory if needed", () => {
    const testDir = ".test_nested_dir" as const;
    const testDbPath = `${testDir}/subdir/data.db`;
    deleteDirIfExists(testDir);

    const config = createDataConfig({ dbPath: testDbPath });
    const result = createDb(config);
    assert.strictEqual(result.ok, true);

    if (result.ok) {
      result.value.close();
    }
    deleteDirIfExists(testDir);
  });

  it("close succeeds", () => {
    const testDbPath = ".test_close.db" as const;
    deleteIfExists(testDbPath);

    const config = createDataConfig({ dbPath: testDbPath });
    const createResult = createDb(config);
    assert.strictEqual(createResult.ok, true);
    if (!createResult.ok) {return;}
    const db = createResult.value;

    const closeResult = db.close();
    assert.strictEqual(closeResult.ok, true);

    deleteIfExists(testDbPath);
  });

  it("schema version is set correctly", () => {
    assert.strictEqual(SCHEMA_VERSION, 1);
  });

  it("error codes are defined", () => {
    assert.strictEqual(ERR_NOT_FOUND, "NOT_FOUND");
    assert.strictEqual(ERR_UNAUTHORIZED, "UNAUTHORIZED");
    assert.strictEqual(ERR_LOCK_HELD, "LOCK_HELD");
    assert.strictEqual(ERR_LOCK_EXPIRED, "LOCK_EXPIRED");
    assert.strictEqual(ERR_VALIDATION, "VALIDATION");
    assert.strictEqual(ERR_DATABASE, "DATABASE");
  });
});
